unittests/fork_db_tests.cpp

```diff
@@ -0,0 +1,107 @@
+#include <eosio/chain/types.hpp>
+#include <eosio/chain/fork_database.hpp>
+#include <eosio/testing/tester.hpp>
+#include <fc/bitutil.hpp>
+#include <boost/test/unit_test.hpp>
+
+
+namespace eosio::chain {
+
+inline block_id_type make_block_id(block_num_type block_num) {
+   static uint32_t nonce = 0;
+   ++nonce;
+   block_id_type id = fc::sha256::hash(std::to_string(block_num) + "-" + std::to_string(nonce));
+   id._hash[0] &= 0xffffffff00000000;
+   id._hash[0] += fc::endian_reverse_u32(block_num); // store the block num in the ID, 160 bits is plenty for the hash
+   return id;
+}
+
+// Used to access privates of block_state
```

Comment:
I found a way to create `block_state_ptr` without accessing private members in my PR :-)

Reply:
Can you point me to what you are suggesting instead?

Reply:
I was referring to [this](https://github.com/AntelopeIO/leap/blob/4aa701fb52214b18ade27ef38503c0785f007156/unittests/finalizer_tests.cpp#L213-L235), but I have not looked at your use in great detail and maybe what I did would not work for your needs.

Reply:
Hmm, I don't think that works. Although, the BSAccessor is really not needed for accessing privates anymore. It is used more as a way to normalize the interface. If we go with two diff `by_best_branch` then `BSAccessor` can be removed.